File: src/desugar.ts
// Clank desugaring pass
// AST-to-AST transform that eliminates syntactic sugar, producing a Core AST
// where the evaluator only handles: literal, var, let, if, apply, lambda, match
//
// Transforms:
//   pipeline:  x |> f(y)      → f(x, y)
//   infix:     a + b          → add(a, b)
//   infix:     a ++ b         → str.cat(a, b)
//   unary:     !x             → not(x)
//   unary:     -x             → negate(x)
//   do-block:  do { x <- e1; e2 } → let x = e1 in e2

import type { Expr, DoStep, MatchArm, HandlerArm, Loc, Pattern, Param } from "./ast.ts";

// Operator-to-function mapping for infix desugaring
const INFIX_OPS: Record<string, string> = {
  "+": "add",
  "-": "sub",
  "*": "mul",
  "/": "div",
  "%": "mod",
  "==": "eq",
  "!=": "neq",
  "<": "lt",
  ">": "gt",
  "<=": "lte",
  ">=": "gte",
  "&&": "and",
  "||": "or",
  "++": "str.cat",
};

const UNARY_OPS: Record<string, string> = {
  "!": "not",
  "-": "negate",
};

export function desugar(expr: Expr): Expr {
  switch (expr.tag) {
    // Core nodes — recurse into children
    case "literal":
      return expr;

    case "var":
      return expr;

    case "let":
      return {
        tag: "let",
        name: expr.name,
        value: desugar(expr.value),
        body: expr.body ? desugar(expr.body) : null,
        loc: expr.loc,
      };

    case "if":
      return {
        tag: "if",
        cond: desugar(expr.cond),
        then: desugar(expr.then),
        else: desugar(expr.else),
        loc: expr.loc,
      };

    case "match":
      return {
        tag: "match",
        subject: desugar(expr.subject),
        arms: expr.arms.map((a: MatchArm) => ({
          pattern: a.pattern,
          body: desugar(a.body),
        })),
        loc: expr.loc,
      };

    case "lambda":
      return {
        tag: "lambda",
        params: expr.params,
        body: desugar(expr.body),
        loc: expr.loc,
      };

    case "apply":
      return {
        tag: "apply",
        fn: desugar(expr.fn),
        args: expr.args.map(desugar),
        loc: expr.loc,
      };

    // Sugar — transform away

    case "pipeline": {
      // x |> f(args...) → f(x, args...)
      // x |> f          → f(x)
      const left = desugar(expr.left);
      const right = desugar(expr.right);
      if (right.tag === "apply") {
        return {
          tag: "apply",
          fn: right.fn,
          args: [left, ...right.args],
          loc: expr.loc,
        };
      }
      // right is a plain function reference: x |> f → f(x)
      return { tag: "apply", fn: right, args: [left], loc: expr.loc };
    }

    case "infix": {
      // Short-circuit: a && b → if a then b else false
      if (expr.op === "&&") {
        return {
          tag: "if",
          cond: desugar(expr.left),
          then: desugar(expr.right),
          else: { tag: "literal", value: { tag: "bool", value: false }, loc: expr.loc },
          loc: expr.loc,
        };
      }
      // Short-circuit: a || b → if a then true else b
      if (expr.op === "||") {
        return {
          tag: "if",
          cond: desugar(expr.left),
          then: { tag: "literal", value: { tag: "bool", value: true }, loc: expr.loc },
          else: desugar(expr.right),
          loc: expr.loc,
        };
      }
      const fn = INFIX_OPS[expr.op] ?? expr.op;
      return {
        tag: "apply",
        fn: { tag: "var", name: fn, loc: expr.loc },
        args: [desugar(expr.left), desugar(expr.right)],
        loc: expr.loc,
      };
    }

    case "unary": {
      const fn = UNARY_OPS[expr.op] ?? expr.op;
      return {
        tag: "apply",
        fn: { tag: "var", name: fn, loc: expr.loc },
        args: [desugar(expr.operand)],
        loc: expr.loc,
      };
    }

    case "do":
      return desugarDo(expr.steps, expr.loc);

    case "for":
      return desugarFor(expr);

    case "range": {
      // Builtin range(start, end) is inclusive [start, end].
      // start..end   (half-open) → range(start, sub(end, 1))
      // start..=end  (inclusive) → range(start, end)
      const start = desugar(expr.start);
      const end = desugar(expr.end);
      const rangeFn: Expr = { tag: "var", name: "range", loc: expr.loc };
      const endArg = expr.inclusive
        ? end
        : {
            tag: "apply" as const,
            fn: { tag: "var" as const, name: "sub", loc: expr.loc },
            args: [end, { tag: "literal" as const, value: { tag: "int" as const, value: 1 }, loc: expr.loc }],
            loc: expr.loc,
          };
      return {
        tag: "apply",
        fn: rangeFn,
        args: [start, endArg],
        loc: expr.loc,
      };
    }

    // Pass-through nodes — recurse but keep the tag
    case "handle":
      return {
        tag: "handle",
        expr: desugar(expr.expr),
        arms: expr.arms.map((a: HandlerArm) => ({
          name: a.name,
          params: a.params,
          resumeName: a.resumeName,
          body: desugar(a.body),
        })),
        loc: expr.loc,
      };

    case "perform":
      return {
        tag: "perform",
        expr: desugar(expr.expr),
        loc: expr.loc,
      };

    case "list":
      return {
        tag: "list",
        elements: expr.elements.map(desugar),
        loc: expr.loc,
      };

    case "tuple":
      return {
        tag: "tuple",
        elements: expr.elements.map(desugar),
        loc: expr.loc,
      };

    case "record":
      return {
        tag: "record",
        fields: expr.fields.map((f: { name: string; value: Expr }) => ({
          name: f.name,
          value: desugar(f.value),
        })),
        loc: expr.loc,
      };

    case "record-update":
      return {
        tag: "record-update",
        base: desugar(expr.base),
        fields: expr.fields.map((f: { name: string; value: Expr }) => ({
          name: f.name,
          value: desugar(f.value),
        })),
        loc: expr.loc,
      };

    case "field-access":
      return {
        tag: "field-access",
        object: desugar(expr.object),
        field: expr.field,
        loc: expr.loc,
      };

    case "borrow":
      return { tag: "borrow", expr: desugar(expr.expr), loc: expr.loc };
    case "clone":
      return { tag: "clone", expr: desugar(expr.expr), loc: expr.loc };
    case "discard":
      return { tag: "discard", expr: desugar(expr.expr), loc: expr.loc };

    default: {
      const _exhaustive: never = expr;
      throw new Error(`Unknown AST node: ${(expr as any).tag}`);
    }
  }
}

/** Flatten do-block steps into nested let expressions */
function desugarDo(steps: DoStep[], loc: Loc): Expr {
  if (steps.length === 0) {
    return { tag: "literal", value: { tag: "unit" }, loc };
  }
  if (steps.length === 1) {
    return desugar(steps[0].expr);
  }
  const [head, ...tail] = steps;
  const rest = desugarDo(tail, loc);
  if (head.bind) {
    return {
      tag: "let",
      name: head.bind,
      value: desugar(head.expr),
      body: rest,
      loc,
    };
  }
  // No binding — sequence: just use a let with a throwaway name
  return {
    tag: "let",
    name: "_",
    value: desugar(head.expr),
    body: rest,
    loc,
  };
}

/** Convert a pattern to lambda params */
function patternToParams(pat: Pattern): Param[] {
  switch (pat.tag) {
    case "p-var":
      return [{ name: pat.name, type: null }];
    case "p-tuple":
      // For tuple destructuring, use a single param name and let match handle it
      // Actually, since lambdas don't support pattern params, we need a wrapper.
      // For simple var patterns, use directly. For complex patterns, use a match.
      return [{ name: "__for_elem", type: null }];
    default:
      return [{ name: "__for_elem", type: null }];
  }
}

/** Wrap body in a match if the pattern is not a simple variable */
function wrapBodyWithPattern(pat: Pattern, body: Expr, loc: Loc): Expr {
  if (pat.tag === "p-var") return body;
  // match __for_elem { pattern => body }
  return {
    tag: "match",
    subject: { tag: "var", name: "__for_elem", loc },
    arms: [{ pattern: pat, body }],
    loc,
  };
}

/**
 * Desugar for-expressions:
 *   for P in E do B            → map(E, fn(P) => B)
 *   for P in E if G do B       → map(filter(E, fn(P) => G), fn(P) => B)
 *   for P in E fold A = I do B → fold(E, I, fn(A, P) => B)
 *   for P in E if G fold A = I do B → fold(filter(E, fn(P) => G), I, fn(A, P) => B)
 */
function desugarFor(expr: Extract<Expr, { tag: "for" }>): Expr {
  const loc = expr.loc;
  const collection = desugar(expr.collection);
  const body = desugar(expr.body);
  const guard = expr.guard ? desugar(expr.guard) : null;

  const elemParams = patternToParams(expr.bind);
  const wrappedBody = wrapBodyWithPattern(expr.bind, body, loc);
  const wrappedGuard = guard ? wrapBodyWithPattern(expr.bind, guard, loc) : null;

  // Build filter(collection, fn(P) => G) if there's a guard
  let source = collection;
  if (wrappedGuard) {
    const filterFn: Expr = {
      tag: "lambda",
      params: elemParams,
      body: wrappedGuard,
      loc,
    };
    source = {
      tag: "apply",
      fn: { tag: "var", name: "filter", loc },
      args: [source, filterFn],
      loc,
    };
  }

  if (expr.fold) {
    // fold form: fold(source, init, fn(acc, P) => B)
    const init = desugar(expr.fold.init);
    const foldFn: Expr = {
      tag: "lambda",
      params: [{ name: expr.fold.acc, type: null }, ...elemParams],
      body: wrappedBody,
      loc,
    };
    return {
      tag: "apply",
      fn: { tag: "var", name: "fold", loc },
      args: [source, init, foldFn],
      loc,
    };
  }

  // map form: map(source, fn(P) => B)
  const mapFn: Expr = {
    tag: "lambda",
    params: elemParams,
    body: wrappedBody,
    loc,
  };
  return {
    tag: "apply",
    fn: { tag: "var", name: "map", loc },
    args: [source, mapFn],
    loc,
  };
}

File: go/internal/desugar/desugar.go
// Package desugar implements the Clank desugaring pass.
//
// AST-to-AST transform that eliminates syntactic sugar, producing a Core AST
// where the evaluator only handles: literal, var, let, if, apply, lambda, match.
//
// Transforms:
//
//	pipeline:  x |> f(y)      → f(x, y)
//	infix:     a + b          → add(a, b)
//	infix:     a ++ b         → str.cat(a, b)
//	unary:     !x             → not(x)
//	unary:     -x             → negate(x)
//	do-block:  do { x <- e1; e2 } → let x = e1 in e2
package desugar

import (
	"fmt"

	"github.com/dalurness/clank/internal/ast"
	"github.com/dalurness/clank/internal/token"
)

// Operator-to-function mapping for infix desugaring.
var infixOps = map[string]string{
	"+":  "add",
	"-":  "sub",
	"*":  "mul",
	"/":  "div",
	"%":  "mod",
	"==": "eq",
	"!=": "neq",
	"<":  "lt",
	">":  "gt",
	"<=": "lte",
	">=": "gte",
	"&&": "and",
	"||": "or",
	"++": "str.cat",
}

var unaryOps = map[string]string{
	"!": "not",
	"-": "negate",
}

// Desugar transforms an expression, eliminating syntactic sugar.
func Desugar(expr ast.Expr) ast.Expr {
	switch e := expr.(type) {
	case ast.ExprLiteral:
		return e
	case ast.ExprVar:
		return e
	case ast.ExprLet:
		var body ast.Expr
		if e.Body != nil {
			body = Desugar(e.Body)
		}
		return ast.ExprLet{Name: e.Name, Value: Desugar(e.Value), Body: body, Loc: e.Loc}
	case ast.ExprLetPattern:
		desugaredValue := Desugar(e.Value)
		var desugaredBody ast.Expr
		if e.Body != nil {
			desugaredBody = Desugar(e.Body)
		} else {
			desugaredBody = ast.ExprLiteral{Value: ast.LitUnit{}, Loc: e.Loc}
		}
		return ast.ExprMatch{
			Subject: desugaredValue,
			Arms:    []ast.MatchArm{{Pattern: e.Pattern, Body: desugaredBody}},
			Loc:     e.Loc,
		}
	case ast.ExprIf:
		return ast.ExprIf{
			Cond: Desugar(e.Cond),
			Then: Desugar(e.Then),
			Else: Desugar(e.Else),
			Loc:  e.Loc,
		}
	case ast.ExprMatch:
		arms := make([]ast.MatchArm, len(e.Arms))
		for i, a := range e.Arms {
			arms[i] = ast.MatchArm{Pattern: a.Pattern, Body: Desugar(a.Body)}
		}
		return ast.ExprMatch{Subject: Desugar(e.Subject), Arms: arms, Loc: e.Loc}
	case ast.ExprLambda:
		return ast.ExprLambda{Params: e.Params, Body: Desugar(e.Body), Loc: e.Loc}
	case ast.ExprApply:
		args := make([]ast.Expr, len(e.Args))
		for i, a := range e.Args {
			args[i] = Desugar(a)
		}
		return ast.ExprApply{Fn: Desugar(e.Fn), Args: args, Loc: e.Loc}

	// Sugar — transform away

	case ast.ExprPipeline:
		left := Desugar(e.Left)
		right := Desugar(e.Right)
		if apply, ok := right.(ast.ExprApply); ok {
			args := make([]ast.Expr, 0, 1+len(apply.Args))
			args = append(args, left)
			args = append(args, apply.Args...)
			return ast.ExprApply{Fn: apply.Fn, Args: args, Loc: e.Loc}
		}
		return ast.ExprApply{Fn: right, Args: []ast.Expr{left}, Loc: e.Loc}

	case ast.ExprInfix:
		// Short-circuit: a && b → if a then b else false
		if e.Op == "&&" {
			return ast.ExprIf{
				Cond: Desugar(e.Left),
				Then: Desugar(e.Right),
				Else: ast.ExprLiteral{Value: ast.LitBool{Value: false}, Loc: e.Loc},
				Loc:  e.Loc,
			}
		}
		// Short-circuit: a || b → if a then true else b
		if e.Op == "||" {
			return ast.ExprIf{
				Cond: Desugar(e.Left),
				Then: ast.ExprLiteral{Value: ast.LitBool{Value: true}, Loc: e.Loc},
				Else: Desugar(e.Right),
				Loc:  e.Loc,
			}
		}
		fn := e.Op
		if mapped, ok := infixOps[e.Op]; ok {
			fn = mapped
		}
		return ast.ExprApply{
			Fn:   ast.ExprVar{Name: fn, Loc: e.Loc},
			Args: []ast.Expr{Desugar(e.Left), Desugar(e.Right)},
			Loc:  e.Loc,
		}

	case ast.ExprUnary:
		fn := e.Op
		if mapped, ok := unaryOps[e.Op]; ok {
			fn = mapped
		}
		return ast.ExprApply{
			Fn:   ast.ExprVar{Name: fn, Loc: e.Loc},
			Args: []ast.Expr{Desugar(e.Operand)},
			Loc:  e.Loc,
		}

	case ast.ExprDo:
		return desugarDo(e.Steps, e.Loc)

	case ast.ExprFor:
		return desugarFor(e)

	case ast.ExprRange:
		start := Desugar(e.Start)
		end := Desugar(e.End)
		rangeFn := ast.ExprVar{Name: "range", Loc: e.Loc}
		var endArg ast.Expr
		if e.Inclusive {
			endArg = end
		} else {
			endArg = ast.ExprApply{
				Fn:   ast.ExprVar{Name: "sub", Loc: e.Loc},
				Args: []ast.Expr{end, ast.ExprLiteral{Value: ast.LitInt{Value: 1}, Loc: e.Loc}},
				Loc:  e.Loc,
			}
		}
		return ast.ExprApply{Fn: rangeFn, Args: []ast.Expr{start, endArg}, Loc: e.Loc}

	// Pass-through nodes — recurse but keep the tag

	case ast.ExprHandle:
		arms := make([]ast.HandlerArm, len(e.Arms))
		for i, a := range e.Arms {
			arms[i] = ast.HandlerArm{
				Name:       a.Name,
				Params:     a.Params,
				ResumeName: a.ResumeName,
				Body:       Desugar(a.Body),
			}
		}
		return ast.ExprHandle{Expr: Desugar(e.Expr), Arms: arms, Loc: e.Loc}

	case ast.ExprPerform:
		return ast.ExprPerform{Expr: Desugar(e.Expr), Loc: e.Loc}

	case ast.ExprList:
		elems := make([]ast.Expr, len(e.Elements))
		for i, el := range e.Elements {
			elems[i] = Desugar(el)
		}
		return ast.ExprList{Elements: elems, Loc: e.Loc}

	case ast.ExprTuple:
		elems := make([]ast.Expr, len(e.Elements))
		for i, el := range e.Elements {
			elems[i] = Desugar(el)
		}
		return ast.ExprTuple{Elements: elems, Loc: e.Loc}

	case ast.ExprRecord:
		fields := make([]ast.RecordField, len(e.Fields))
		for i, f := range e.Fields {
			fields[i] = ast.RecordField{Name: f.Name, Tags: f.Tags, Value: Desugar(f.Value)}
		}
		var spread ast.Expr
		if e.Spread != nil {
			spread = Desugar(e.Spread)
		}
		return ast.ExprRecord{Fields: fields, Spread: spread, Loc: e.Loc}

	case ast.ExprRecordUpdate:
		fields := make([]ast.RecordUpdateField, len(e.Fields))
		for i, f := range e.Fields {
			fields[i] = ast.RecordUpdateField{Name: f.Name, Value: Desugar(f.Value)}
		}
		return ast.ExprRecordUpdate{Base: Desugar(e.Base), Fields: fields, Loc: e.Loc}

	case ast.ExprFieldAccess:
		return ast.ExprFieldAccess{Object: Desugar(e.Object), Field: e.Field, Loc: e.Loc}

	case ast.ExprBorrow:
		return ast.ExprBorrow{Expr: Desugar(e.Expr), Loc: e.Loc}
	case ast.ExprClone:
		return ast.ExprClone{Expr: Desugar(e.Expr), Loc: e.Loc}
	case ast.ExprDiscard:
		return ast.ExprDiscard{Expr: Desugar(e.Expr), Loc: e.Loc}

	default:
		panic(fmt.Sprintf("desugar: unknown AST node type %T", expr))
	}
}

// desugarDo flattens do-block steps into nested let expressions.
func desugarDo(steps []ast.DoStep, loc token.Loc) ast.Expr {
	if len(steps) == 0 {
		return ast.ExprLiteral{Value: ast.LitUnit{}, Loc: loc}
	}
	if len(steps) == 1 {
		return Desugar(steps[0].Expr)
	}
	head := steps[0]
	rest := desugarDo(steps[1:], loc)

	// If the step is a bare `let x = e` (no `in`, no `<-` bind), lift the
	// let so the remainder of the block becomes its body. This makes
	//   do { let x = e1; e2 }
	// equivalent to
	//   do { x <- e1; e2 }
	// i.e. `x` is in scope for all subsequent steps.
	if head.Bind == "" {
		if letExpr, ok := head.Expr.(ast.ExprLet); ok && letExpr.Body == nil {
			return ast.ExprLet{
				Name:  letExpr.Name,
				Value: Desugar(letExpr.Value),
				Body:  rest,
				Loc:   letExpr.Loc,
			}
		}
	}

	name := head.Bind
	if name == "" {
		name = "_"
	}
	return ast.ExprLet{
		Name:  name,
		Value: Desugar(head.Expr),
		Body:  rest,
		Loc:   loc,
	}
}

// patternToParams converts a pattern to lambda params.
func patternToParams(pat ast.Pattern) []ast.Param {
	if pv, ok := pat.(ast.PatVar); ok {
		return []ast.Param{{Name: pv.Name}}
	}
	return []ast.Param{{Name: "__for_elem"}}
}

// wrapBodyWithPattern wraps body in a match if the pattern is not a simple variable.
func wrapBodyWithPattern(pat ast.Pattern, body ast.Expr, loc token.Loc) ast.Expr {
	if _, ok := pat.(ast.PatVar); ok {
		return body
	}
	return ast.ExprMatch{
		Subject: ast.ExprVar{Name: "__for_elem", Loc: loc},
		Arms:    []ast.MatchArm{{Pattern: pat, Body: body}},
		Loc:     loc,
	}
}

// desugarFor desugars for-expressions using runtime-dispatched builtins.
func desugarFor(expr ast.ExprFor) ast.Expr {
	loc := expr.Loc
	collection := Desugar(expr.Collection)
	body := Desugar(expr.Body)
	var guard ast.Expr
	if expr.Guard != nil {
		guard = Desugar(expr.Guard)
	}

	elemParams := patternToParams(expr.Bind)
	wrappedBody := wrapBodyWithPattern(expr.Bind, body, loc)
	var wrappedGuard ast.Expr
	if guard != nil {
		wrappedGuard = wrapBodyWithPattern(expr.Bind, guard, loc)
	}

	source := collection
	if wrappedGuard != nil {
		filterFn := ast.ExprLambda{Params: elemParams, Body: wrappedGuard, Loc: loc}
		source = ast.ExprApply{
			Fn:   ast.ExprVar{Name: "__for_filter", Loc: loc},
			Args: []ast.Expr{source, filterFn},
			Loc:  loc,
		}
	}

	if expr.Fold != nil {
		init := Desugar(expr.Fold.Init)
		foldParams := make([]ast.Param, 0, 1+len(elemParams))
		foldParams = append(foldParams, ast.Param{Name: expr.Fold.Acc})
		foldParams = append(foldParams, elemParams...)
		foldFn := ast.ExprLambda{Params: foldParams, Body: wrappedBody, Loc: loc}
		return ast.ExprApply{
			Fn:   ast.ExprVar{Name: "__for_fold", Loc: loc},
			Args: []ast.Expr{source, init, foldFn},
			Loc:  loc,
		}
	}

	mapFn := ast.ExprLambda{Params: elemParams, Body: wrappedBody, Loc: loc}
	return ast.ExprApply{
		Fn:   ast.ExprVar{Name: "__for_each", Loc: loc},
		Args: []ast.Expr{source, mapFn},
		Loc:  loc,
	}
}
